File: ios/RCTImageMarker/MarkWatermarkOptions.swift
//
//  MarkWatermarkOptions.swift
//  react-native-image-marker
//

import Foundation
import UIKit
import React

enum WatermarkLayerOptions {
    case text(TextOptions)
    case image(WatermarkImageOptions)

    init(dicOpts opts: [AnyHashable: Any]) throws {
        guard let type = opts["type"] as? String else {
            throw NSError(domain: ErrorDomainEnum.PARAMS_REQUIRED.rawValue, code: 0, userInfo: [NSLocalizedDescriptionKey: "watermark layer type is required"])
        }
        switch type {
        case "text":
            self = .text(try TextOptions(dicOpts: opts))
        case "image":
            self = .image(try WatermarkImageOptions(dicOpts: opts))
        default:
            throw NSError(domain: ErrorDomainEnum.PARAMS_INVALID.rawValue, code: 0, userInfo: [NSLocalizedDescriptionKey: "watermark layer type is invalid"])
        }
    }
}

class MarkWatermarkOptions: Options {
    var watermarkLayers: [WatermarkLayerOptions] = []
    var imageLayers: [WatermarkImageOptions] {
        watermarkLayers.compactMap { layer in
            if case let .image(imageOptions) = layer {
                return imageOptions
            }
            return nil
        }
    }

    override init(dicOpts opts: [AnyHashable: Any]) throws {
        try super.init(dicOpts: opts)
        guard let watermarkOpts = opts["watermarks"] as? [[AnyHashable: Any]], watermarkOpts.count > 0 else {
            throw NSError(domain: ErrorDomainEnum.PARAMS_REQUIRED.rawValue, code: 0, userInfo: [NSLocalizedDescriptionKey: "watermarks is required"])
        }
        self.watermarkLayers = try watermarkOpts.map { try WatermarkLayerOptions(dicOpts: $0) }
    }

    static func checkWatermarkParams(_ opts: [AnyHashable: Any], rejecter reject: @escaping RCTPromiseRejectBlock) -> MarkWatermarkOptions? {
        do {
            return try MarkWatermarkOptions(dicOpts: opts)
        } catch let error as NSError {
            reject(error.domain, error.localizedDescription, nil)
            return nil
        }
    }
}

File: android/src/main/java/com/jimmydaddy/imagemarker/base/MarkWatermarkOptions.kt
package com.jimmydaddy.imagemarker.base

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap

sealed class WatermarkLayerOptions {
  data class TextLayer(val textOptions: TextOptions) : WatermarkLayerOptions()
  data class ImageLayer(val imageOptions: WatermarkImageOptions) : WatermarkLayerOptions()

  companion object {
    fun fromReadableMap(options: ReadableMap?): WatermarkLayerOptions {
      if (options == null) {
        throw MarkerError(ErrorCode.NULL_MAP, "watermark layer is null")
      }
      val type = options.getString("type")
      return when (type) {
        "text" -> TextLayer(TextOptions(options))
        "image" -> ImageLayer(WatermarkImageOptions(options))
        else -> throw MarkerError(
          ErrorCode.INVALID_PARAMS,
          "watermark layer type is required"
        )
      }
    }
  }
}

class MarkWatermarkOptions(options: ReadableMap) : Options(options) {
  val watermarkLayers: Array<WatermarkLayerOptions>
  val imageLayers: List<WatermarkLayerOptions.ImageLayer>

  init {
    val watermarkOpts = options.getArray("watermarks")
    if (watermarkOpts == null || watermarkOpts.size() <= 0) {
      throw MarkerError(
        ErrorCode.PARAMS_REQUIRED,
        "watermarks is required"
      )
    }

    val layers = arrayListOf<WatermarkLayerOptions>()
    for (i in 0 until watermarkOpts.size()) {
      layers.add(WatermarkLayerOptions.fromReadableMap(watermarkOpts.getMap(i)))
    }
    watermarkLayers = layers.toTypedArray()
    imageLayers = layers.filterIsInstance<WatermarkLayerOptions.ImageLayer>()
  }

  companion object {
    @JvmStatic
    fun checkParams(opts: ReadableMap, promise: Promise): MarkWatermarkOptions? {
      try {
        return MarkWatermarkOptions(opts)
      } catch (e: Exception) {
        val markerError = MarkerError.fromInvalidParams(e, "Invalid watermark options")
        promise.reject(markerError.getErrorCode(), markerError.getErrMsg())
      }
      return null
    }
  }
}
